scalabel/eval/tagging.py

```diff
@@ -17,6 +32,434 @@
+        tp = y_true == y_pred
+        tp_bins = y_true[tp]
+        tp_bins_weights = None
+
+        true_sum = pred_sum = tp_sum = np.zeros(len(labels))
+        if len(tp_bins):
+            tp_sum = np.bincount(
+                tp_bins, weights=tp_bins_weights, minlength=len(labels)
+            )
+
+        if len(y_pred):
+            pred_sum = np.bincount(y_pred, minlength=len(labels))
+        if len(y_true):
+            true_sum = np.bincount(y_true, minlength=len(labels))
+
+        # retain only selected labels
+        indices = np.searchsorted(sorted_labels, labels[:n_labels])
+        tp_sum = tp_sum[indices]
+        true_sum = true_sum[indices]
+        pred_sum = pred_sum[indices]
+
+    else:
+        sum_axis = 0
+        if n_labels is not None:
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
+        # calculate weighted counts
+        true_and_pred = y_true.multiply(y_pred)
+        tp_sum = _count_nonzero(true_and_pred, axis=sum_axis)
+        pred_sum = _count_nonzero(y_pred, axis=sum_axis)
+        true_sum = _count_nonzero(y_true, axis=sum_axis)
+
+    fp = pred_sum - tp_sum
+    fn = true_sum - tp_sum
+    tp = tp_sum
+
+    tn = y_true.shape[0] - tp - fp - fn
+    return np.array([tn, fp, fn, tp]).T.reshape(-1, 2, 2)
+
+
+def _prf_divide(numerator, denominator):
+    """Performs division and handles divide-by-zero."""
+    mask = denominator == 0.0
+    denominator = denominator.copy()
+    denominator[mask] = 1  # avoid infs/nans
+    result = numerator / denominator
+
+    if not np.any(mask):
+        return result
+
+    # address zero division by setting 0s to 1s
+    result[mask] = 1.0
+    return result
+
+
+def _precision_recall_fscore_support(
+    y_true: NDArrayI32,
+    y_pred: NDArrayI32,
+    labels: NDArrayI32,
+    average: Optional[str] = None,
+    beta: float = 1.0,
+) -> Tuple[np.float64, np.float64, np.float64, Optional[NDArrayI32]]:
+    """Compute precision, recall, F-measure and support for each class."""
+    mcm = _multilabel_confusion_matrix(y_true, y_pred, labels)
+    tp_sum: NDArrayI32 = mcm[:, 1, 1]
+    pred_sum = tp_sum + mcm[:, 0, 1]
+    true_sum: Optional[NDArrayI32] = tp_sum + mcm[:, 1, 0]
+
+    if average == "micro":
+        tp_sum = np.array([tp_sum.sum()])
+        pred_sum = np.array([pred_sum.sum()])
+        assert true_sum is not None
+        true_sum = np.array([true_sum.sum()])
+
+    beta2 = beta ** 2
+
+    # divide and set scores
+    precision = _prf_divide(tp_sum, pred_sum)
+    recall = _prf_divide(tp_sum, true_sum)
+
+    # if tp == 0 F will be 1 only if all predictions are zero, all labels are
+    # zero, and zero_division=1. In all other case, 0
+    if np.isposinf(beta):
+        f_score = recall
+    else:
+        denom = beta2 * precision + recall
+
+        denom[denom == 0.0] = 1  # avoid division by 0
+        f_score = (1 + beta2) * precision * recall / denom
+
+    if average == "weighted":
```

Comment:
Is this ever used? I dint remember using this

Reply:
I'll take it out